middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -1465,84 +1522,145 @@ let close_let_rec acc env ~function_declarations
+  let fresh_group_id =
+    let i = ref 0 in
+    fun () ->
+      let n = !i in
+      incr i;
+      n
+  in
+  let group_to_bound_consts, symbol_to_groups =
+    Code_id.Map.fold
+      (fun code_id code (g2c, s2g) ->
+        let id = fresh_group_id () in
+        let bound = Bound_symbols.Pattern.code code_id in
+        let const = Static_const_or_code.create_code code in
+        ( GroupMap.add id (bound, const) g2c,
+          CIS.Map.add (CIS.create_code_id code_id) id s2g ))
       all_code
+      (GroupMap.empty, CIS.Map.empty)
+  in
+  let group_to_bound_consts, symbol_to_groups =
+    List.fold_left
+      (fun (g2c, s2g) (symbols, set_of_closures) ->
+        let id = fresh_group_id () in
+        let symbols = Closure_id.Lmap.map fst symbols in
+        let bound = Bound_symbols.Pattern.set_of_closures symbols in
+        let const =
+          Static_const_or_code.create_static_const
+            (Set_of_closures set_of_closures)
+        in
+        ( GroupMap.add id (bound, const) g2c,
+          Closure_id.Lmap.fold
+            (fun _closure_id symbol s2g ->
+              CIS.Map.add (CIS.create_symbol symbol) id s2g)
+            symbols s2g ))
+      (group_to_bound_consts, symbol_to_groups)
+      sets_of_closures
+  in
+  let graph =
+    GroupMap.map
+      (fun (_bound, const) ->
+        let free_names = Static_const_or_code.free_names const in
+        let deps =
+          Code_id.Set.fold
+            (fun code_id deps ->
+              match
+                CIS.Map.find (CIS.create_code_id code_id) symbol_to_groups
+              with
+              | exception Not_found -> deps
+              | id -> Numbers.Int.Set.add id deps)
+            (Name_occurrences.code_ids free_names)
+            Numbers.Int.Set.empty
+        in
+        Symbol.Set.fold
+          (fun symbol deps ->
+            match CIS.Map.find (CIS.create_symbol symbol) symbol_to_groups with
+            | exception Not_found -> deps
+            | id -> Numbers.Int.Set.add id deps)
+          (Name_occurrences.symbols free_names)
+          deps)
+      group_to_bound_consts
   in
   let components = SCC.connected_components_sorted_from_roots_to_leaf graph in
   Array.fold_left
     (fun (acc, body) (component : SCC.component) ->
-      let code_ids =
+      let group_ids =
         match component with
-        | No_loop code_id -> [code_id]
-        | Has_loop code_ids -> code_ids
+        | No_loop group_id -> [group_id]
+        | Has_loop group_ids -> group_ids
       in
       let bound_symbols, static_consts =
         List.fold_left
-          (fun (bound_symbols, static_consts) code_id ->
-            let bound_symbols =
-              Bound_symbols.Pattern.code code_id :: bound_symbols
-            in
-            let code =
-              try Code_id.Map.find code_id all_code
+          (fun (bound_symbols, static_consts) group_id ->
+            let bound_symbol, static_const =
```

Comment:
Despite the singular used in `bound_symbol`, this is a `Bound_symbol.Pattern.t` which can include several symbols (if binding a set of closures). So despite the fact that currently we don't generate static sets of closures with more than one closure in Closure_conversion, I'd prefer if we used a different name like `bound_pattern`.

Reply:
The code never dissociates `Lletrec` containing several functions. It is lifted as it is as long as there is no closure variables.

Reply:
Ah, indeed. It tracks the approximations and substitutions for individual symbols, but it does lift a single set of closures.
We could actually split the set of closures into individual closures (and it might be actually helpful), but we can reconsider this later.